### PR TITLE
feat: collection runner

### PR DESCRIPTION
**Pull request by anwarulislam**
Closes HFE-236

### Description

This PR introduces the `Collection Runner`, a new feature that allows users to run API collections and manage their execution flow.

### Key Features:

* **Personal & Team Collections:** Run both personal and team-shared collections, enhancing collaboration and consistency.
* **Environment Modification:** Modify environment variables during execution, enabling flexible and context-specific testing.
* **Nested Collections & Folders:** Organize collections into nested folders, facilitating structure and maintainability.
* **Advanced Configuration:** Provides basic support for advanced configuration options, including `stopOnError` functionality. Further enhancements for features like `persistResponse` are planned for future releases.

**Notes:**

* The implementation of advanced configuration options is currently limited. `stopOnError` is functional, but `persistResponse` requires further development.

- [ ] Not Completed
- [x] Completed

### Images

image
image

### Checks

- [x] My pull request adheres to the code style of this project
- [x] My code requires changes to the documentation
- [ ] I have updated the documentation as required
- [ ] All the tests have passed

### CODE REVIEW

**Comment by tomerr-ma** [comment]
Any chance this PR will be merged soon?

**Comment by LewisBango** [comment]
@anwarulislam hi, is there any update on this one please?

**Comment by guysha20** [comment]
Hi @anwarulislam, any time estimation for merging this one?

**Comment by MarcelHoell** [comment]
Hi, this is gonna be a very nice feature! When to you plan to release it?

**Comment by amishratnasthapit** [comment]
Looking forward to this feature.

**Comment by christopher-nash** [comment]
I noticed the docs have a page which mentions the runner: https://docs.hoppscotch.io/documentation/features/runner but apparently those documents were updated prematurely?

The lack of a runner is the one feature thats holding me back from entirely jumping ship from Postman so I'm really eager for this feature.

**Review by marcelocastilho — COMMENTED** [review]
:eyes:

**Comment by Zamion101** [comment]
This feature is a really wanted feature for to be able to test REST APIs quickly and autonomously. Last update/commit in this PR dates to 3 months ago and no further update mentioned or PR closed. If its possible we would like to get some update on this PR and ETA of this update to main branch. Thanks.

**Comment by thetronjohnson** [comment]
Hi, quick update on the collection runner: we're waiting for a refactor to be finished for the collection runner to work at its best. We're focusing on completing the refactor this month and getting this feature merged.

**Comment by LewisBango** [comment]
@thetronjohnson Thanks, that's great to hear.

One question, will the runner support running individual folders as well as the entire collection?

**Comment by RJRReid** [comment]
I work for a large company that very soon will need to replace POSTMAN.
However, whatever it gets replaced with will have to have a Runner that uses a CSV as input.
If you can get this working, I might convince my company to purchase several hundred of the licenses for the paid version.

**Comment by WanwanNoOne** [comment]
@anwarulislam hi, is there any update on this one please?

**Comment by WanwanNoOne** [comment]
@hoppscotch hi, is there any update on this one please?

**Comment by NeilRiver** [comment]
apparently never

⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⣀⣠⣤⣴⣶⣶⣶⣶⣶⣶⣶⣶⣶⣶⣶⣦⣤⣀⡀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⣀⣴⣾⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣶⣶⣤⣀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⣠⣾⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣷⣤⡀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⢀⣤⣾⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣦⡀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⢀⣾⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣷⣄⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⣼⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣆⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⢸⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⡿⠿⠿⠿⣿⠿⠿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣦⣄⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⢠⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⠛⠋⢻⣇⠀⠀⠸⣇⠀⠀⢰⡏⠀⣸⠇⠀⢀⡿⠀⢹⡏⢹⡟⠻⣿⣿⣿⣦⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⣸⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⡿⠿⣯⡁⠹⣧⣀⣤⣿⣦⢴⣶⣿⣶⣶⣿⠥⠤⠿⠦⠤⢼⣧⣀⣿⣀⣼⠃⢰⡏⠈⣿⢿⡆⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⢀⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⡟⠉⢷⡄⠘⣷⣶⠿⠽⠒⠚⠉⠉⠀⠀⠀⠀⠉⠙⠒⠒⠒⠒⠒⠒⠒⢼⠉⠙⠓⢿⣄⣰⡿⢸⣷⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⢸⣿⣿⣿⣿⣿⣿⣿⣿⣿⡛⠉⠻⣆⣀⣿⠟⠉⠀⠀⠀⠀⣀⣀⣀⣀⣀⣀⠀⠀⠀⠀⠀⠀⠀⠀⠀⣀⣀⣀⡀⠀⠀⠈⠻⢷⡾⣿⡇⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⢸⣿⣿⣿⣿⣿⣿⡿⣯⡈⠻⢶⣤⠾⠋⠁⠀⠀⠴⠚⠉⠉⠁⠀⠀⠀⠀⢩⡀⠀⠀⠀⠀⠀⠀⠐⡋⠀⠀⠀⠉⠳⣄⠀⠀⠘⣷⣿⡇⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⢸⣿⣿⣿⣿⣿⠻⣦⡈⢻⣴⠟⠁⠀⠀⠀⠀⠀⠀⣤⠤⠤⠴⠖⠀⠀⠀⠀⠹⡆⠀⠀⠀⠀⠀⠀⣿⠀⠀⠀⣀⣀⠀⠀⠀⠀⣿⣾⠁⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⢸⣿⣿⣿⡋⠙⢦⣌⣻⡿⠁⠀⠀⠀⠀⠀⠀⡀⠀⠀⠀⣀⣀⣀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠛⠀⠀⠀⠀⠈⠉⠓⠦⠄⣸⠇⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⢸⣿⣿⠉⠻⣦⡀⢹⡟⠀⠀⣀⠀⠀⠀⠀⠼⠿⠗⠸⣯⡉⣿⣿⣿⣿⠟⡇⠀⠀⠀⠀⠀⠀⠀⠀⠀⣴⠶⢶⣤⣤⣤⠀⢀⡐⣿⠀⠀⠀⠀⠀⠀⠀⡀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠸⣿⠙⣧⡀⠉⣻⡟⠁⠀⠀⢱⣧⣤⠀⠀⠀⠀⠀⢰⠾⠿⡿⠿⠿⣿⣿⣷⠀⠀⠀⠀⠀⠀⠀⠀⠰⣟⢶⣾⣿⣿⣿⠀⠻⣤⣿⡀⠀⠀⠀⠀⠀⢰⠃⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠹⣆⠈⠻⣦⡟⠀⠀⠀⠀⢘⣷⢿⠄⠀⠀⠀⠀⠘⠓⠲⠿⣿⣿⣷⡿⠃⠀⠀⠀⠀⠀⠀⠀⣷⠀⠙⠿⣷⣾⣿⣏⣤⠀⠀⢸⣷⠀⠀⠀⠀⠀⣸⠁⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠹⣆⠀⢸⡇⠀⠀⠀⠀⢰⣿⣬⡳⣃⡄⠀⠀⠀⠀⠀⠀⠈⠉⠁⠀⠀⠀⠀⠀⠀⠀⠀⠀⢻⣆⠀⠀⠙⠃⠀⠀⠀⠀⠀⠚⣿⠀⠀⠀⠀⣠⣾⠏⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠙⢷⣼⠇⠀⠀⠀⠀⢸⡇⢬⠱⢸⣘⣦⡀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠻⣦⡀⠀⠀⠀⢀⡈⢀⠰⣝⣿⠀⠀⢰⣳⠋⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⣙⣧⠀⠀⠀⠀⠸⣷⣃⢯⡤⡿⣅⠀⢦⡀⠀⠀⠀⠀⠀⠀⣠⡤⠶⠀⠀⠀⠀⠀⠀⠀⠈⣿⠀⠀⠀⠀⠑⢿⡟⣺⠏⠀⠀⣧⠃⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠸⢿⡇⠀⠀⠀⠀⢹⣗⡃⠹⢄⠙⣆⠤⠘⠀⠀⠀⠀⢠⡀⢻⣆⠀⠀⢀⣀⡀⠀⠀⠀⣾⠃⠀⠀⠰⢠⢚⣿⣿⡟⠀⠀⢹⡀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⣾⡇⠀⠀⠀⠀⠈⢻⣇⠞⣐⣤⠩⢹⠂⠀⠲⠀⢠⡞⢁⡀⣉⠉⠁⠨⠀⢀⠤⠉⠛⠃⢾⠀⠠⠴⣳⢉⣻⡟⠀⠀⢠⣷⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⣼⡇⠀⠀⠀⠀⠀⣆⢻⣾⣣⠈⠁⡄⠒⢿⠉⢠⠏⠀⢶⠠⡴⠛⠃⢑⣽⡩⢴⣬⣅⣀⠸⡆⠈⢼⢺⣼⠏⠀⠀⠀⢨⡿⡇⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠰⣿⡇⠀⠀⠀⠀⠀⢸⠈⡛⢻⣶⣄⡼⡚⢤⠁⠛⠀⠀⠾⠳⠦⢤⣄⣀⣀⣀⣀⣤⣯⣽⡳⢿⡀⢖⣻⠋⠀⠀⠀⠀⢀⡇⣧⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⣀⣤⣤⣴⣾⣿⠁⠀⠀⠀⠀⠀⣿⠨⠀⠀⠘⠻⣿⣾⡌⠓⣧⡄⠀⠀⡀⠀⠀⠰⢦⠶⡦⠤⠾⢭⠳⢿⡲⣔⣶⣿⠀⠀⠀⠀⠀⣸⣣⡟⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⣀⣠⣶⣾⣿⣿⣿⣿⣿⣿⠃⠀⠀⠀⠀⠀⣘⡋⣀⢃⡡⠙⠦⠰⠙⣿⣄⡌⠩⠁⣄⠐⠢⡌⠖⢀⢀⡤⠄⡠⣀⢰⠤⢹⣷⣽⡫⣄⠀⣠⢔⣫⠴⠋⠀⠀⠀⠀⠀⠀
⠀⠀⠀⣰⣾⣿⣿⣿⣿⣿⣿⣿⣿⣿⡿⠀⠀⠀⠀⠀⠀⠸⣍⣒⡉⠊⣩⠦⠓⢀⠘⠹⣿⣧⡁⡶⠀⠀⠹⠀⡴⠂⠀⠋⢰⠒⠘⣤⣼⡋⠀⠉⠲⡿⠛⠉⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⣠⣶⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⡿⠁⠀⠀⠀⠀⠀⣀⣤⡶⠉⣂⠒⠁⠀⢈⡂⢄⠀⠈⠸⡻⡷⢦⣤⣸⣁⣀⣈⣄⣀⣠⣴⣾⣿⣿⣿⣦⡀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⣾⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣥⣀⠀⠀⠀⠀⣾⠋⠀⠀⠀⠀⠀⠙⠀⠱⠝⢢⠉⠙⠆⠑⠟⣇⠠⠉⠉⠈⢙⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣷⣄⡀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣷⣾⡁⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠈⠀⠈⠢⡄⠀⠘⠧⣄⠀⠀⠀⠙⠿⠿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣦⡀⠀⠀⠀⠀⠀⠀⠀⠀
⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣷⣦⣀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠘⠀⠁⠀⠈⠀⠁⢀⣠⣴⣾⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣷⣶⣶⣶⣦⣤⣀⠀⠀
⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⡶⠦⢤⣴⣖⣲⣄⣀⡀⢀⣠⠤⢤⣀⣤⣴⣾⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣶
⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣄⢘⠳⡶⠾⠿⡿⠿⠿⡭⢉⣩⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿

**Comment by m00ncr4wler** [comment]
Awesome job so far 👍

**Comment by amishratnasthapit** [comment]
Really looking forward to this feature.

File: packages/hoppscotch-selfhost-web/src/platform/collections/collections.platform.ts [hunk 1]
```diff
@@ -70,13 +70,14 @@ function initCollectionsSync() {
...
-  loadUserCollections("REST")
+  // TODO: fix collection schema transformation on backend maybe?
+  // loadUserCollections("REST")
   loadUserCollections("GQL")
 
   // TODO: test & make sure the auth thing is working properly
   currentUser$.subscribe(async (user) => {
     if (user) {
-      loadUserCollections("REST")
+      // loadUserCollections("REST")
```
Comment (jamesgeorge007) [inline packages/hoppscotch-selfhost-web/src/platform/collections/collections.platform.ts:80]: Could you elaborate on why `loadUserCollections` is commented out? Is anything remaining here with the TODO resolution?

**Review by jamesgeorge007 — CHANGES_REQUESTED** [review]
Great work, leaving a couple of edge cases/UX improvements encountered while testing the flows below:

- Attempting to run an empty collection should be prevented. An error toast is displayed for root collections under personal workspace though not consistent, but not extended to child collections. At the same time, it has to be accounted for collections at all levels under shared workspaces.

  There's also an edge case with root collections under the personal workspace where the error toast is still displayed despite creating requests within it at times.

  image

- Say, a collection run is initiated. Now, if the requests get deleted resulting in an empty collection, a corresponding empty state would be nice to have in the test runner tab open.

  image

- If the collection corresponding to the test runner tab gets deleted there is no feedback currently and the below error gets logged to the console:

  image

- We could employ empty states for the `Passed` & `Failed` tabs when no tests exist in the requests getting executed.

  image

- If a request has an invalid endpoint, issues with scripts, running via the interceptor set to `Proxy` outside hoppscotch.io, etc, that could lead to an error running the collection, the corresponding request error state appears across the `Passed` and `Failed` tabs.

  image

- Consider a collection with a request within it. Adding requests under the above collection would open the same in a new tab. Now, clicking on the same request will keep opening it in multiple tabs. This is observed for requests being added after initiating a collection run.

- In the test run tab, only the immediate parent collection name is displayed although there's a visual cue regarding the overall hierarchy. It'd be nice to display the collection tree or convey the full path to the request executed.

  image

- There's a schema mismatch reported with a test run tab open against the key `restTabState` and performing a page reload. At times, the above tab gets closed during page reload.

- The response selected for any request from the previous collection run persists while initiating another run.

  image

- Specifying a slightly large delay in the run configuration and attempting to stop the run in between it is observed that the intermediary request stays in an infinite loading state.

  image

- Running a collection with authorization and headers specified at the collection level isn't inherited by requests as expected. Try importing [this collection](https://github.com/anwarulislam/hoppscotch/blob/main/packages/hoppscotch-cli/src/__tests__/e2e/fixtures/collections/collection-level-auth-headers-coll.json) and observe the requests succeed while executing individually and not during the collection run.

- While selecting `New Run`, persisting the active `Delay` value might be helpful with a fallback to `500ms`. Also, entering a negative or decimal value for the `Delay` field should be prevented.

- Checking `Stop run if an error occurs` under run configuration and in case a request has syntax errors with the post-request script, the execution isn't halted and proceeds to completion. Also, do we want to report such a case under `Failures`, or keep it exclusive to failed test cases?

  image

- Say, there's a typo in the post request script for a request with the `toBeLevel()` scripting API method usage, it isn't reported as a failure. This is a pre-existing issue with the scripting side.

```js
// Check status code is 3xx
pw.test("Status code is 3xx", ()=> {
    // `status` instead of `status`
    pw.expect(pw.response.statu).toBeLevel3xx();
});
```

    image

- Is it intentional that having a test runner tab open and attempting a new run from the collection tree opens another tab instead of picking the existing one?

I've pushed a [quick change](https://github.com/hoppscotch/hoppscotch/pull/3600/commits/679e809ea6a6e05d44ebe49f5cf26f082bbc3dbe) resolving schema mismatch errors with personal collections. Also, please look into the failing CLI tests.

**Review by jamesgeorge007 — APPROVED** [review]
The remaining issues are to be handled separately.

**Review by nivedin — APPROVED** [review]
LGTM 💯

**Comment by jamesgeorge007** [comment]
Collection runner is now live with [v2024.11.0](https://github.com/hoppscotch/hoppscotch/releases/tag/2024.11.0).

**Pull request merged**